File: Documentation/5.4/Samples/python/ClientApi/Session/Querying/TextSearch/UsingRegex.py
from typing import TypeVar

from ravendb import DocumentQuery

from examples_base import ExampleBase, Product

_T = TypeVar("_T")


class UsingRegex(ExampleBase):
    def setUp(self):
        super().setUp()

    def test_example(self):
        with self.embedded_server.get_document_store("Regex") as store:
            with store.open_session() as session:
                # region regex_1
                # loads all products, which name
                # starts with 'N' or 'A'
                products = list(session.query(object_type=Product).where_regex("Name", "^[NA]"))
                # endregion


class Foo:
    # region syntax
    def where_regex(self, field_name: str, pattern: str) -> DocumentQuery[_T]: ...

    # endregion

File: Documentation/5.4/Samples/php/ClientApi/Session/Querying/TextSearch/UsingRegex.php
<?php

use RavenDB\Documents\Session\FilterDocumentQueryBaseInterface;
use RavenDB\Samples\Infrastructure\DocumentStoreHolder;
use RavenDB\Samples\Infrastructure\Orders\Product;

interface FooInterface
{
    # region syntax
    function whereRegex(?string $fieldName, ?string $pattern): FilterDocumentQueryBaseInterface;
    # endregion
}

class UsingRegex
{
    public function examples(): void
    {
        $store = DocumentStoreHolder::getStore();
        try {
            $session = $store->openSession();
            try {
                # region regex_1
                /** @var array<Product> $products */
                $products = $session
                    ->query(Product::class)
                    ->whereRegex("Name", "^[NA]")
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

        } finally {
            $store->close();
        }
    }
}

class User
{
    public ?string $firstName = null;

    public function getFirstName(): ?string
    {
        return $this->firstName;
    }

    public function setFirstName(?string $firstName): void
    {
        $this->firstName = $firstName;
    }
}
